heat/core/operations.py

```diff
@@ -544,7 +549,620 @@ def sum(x, axis=None, out=None):
 
     return __reduce_op(x, torch.sum, MPI.SUM, axis, out)
 
-  
+
+def merge_means(mu1, n1, mu2, n2):
+    """
+    Function to merge two means by pairwise update
+
+    Parameters
+    ----------
+    mu1 : 1D ht.tensor or 1D torch.tensor
+          Calculated mean
+    n1 : 1D ht.tensor or 1D torch.tensor
+         number of elements used to calculate mu1
+    mu2 : 1D ht.tensor or 1D torch.tensor
+          Calculated mean
+    n2 : 1D ht.tensor or 1D torch.tensor
+         number of elements used to calculate mu2
+
+    Returns
+    -------
+    mean of combined set
+    number of elements in the combined set
+    """
+    if mu1.item() == np.nan:
+        mu1 = 0.0
+    if mu2.item() == np.nan:
+        mu2 = 0.0
+    delta = mu2.item() - mu1.item()
+    n1 = n1.item()
+    n2 = n2.item()
+    return mu1 + n2 * (delta / (n1+n2)), n1 + n2
+
+
+def mean(x, dimen=None, all_procs=False):
+    """
+    Calculates and returns the mean of a tensor.
+    If a dimension is given, the mean will be taken in that direction.
+
+    Parameters
+    ----------
+    x : ht.tensor
+        Values for which the mean is calculated for
+    dimen : None, Int, iterable
+            Dimension which the mean is taken in.
+            Default: None -> mean of all data calculated
+    all_procs : Bool
+                Flag to distribute the data to all processes
+                If True: will split the result in the same direction as x
+                Default: False (mean of the whole dataset still calculated but not on every node)
+
+    Returns
+    -------
+    ht.tensor containing the mean/s, if split, then split in the same direction as x.
+    """
+
+    def combine_all_means(target_dimens):
+        """
+        Function to *combine* all the calculated means across the processes.
+        This function will only be used if the all_procs flag is true.
+        This function will operate on x from the mean function.
+
+        Parameters
+        ----------
+        target_dimens : iterable
+                        iterable with the dimensions of the output of the mean function
+
+        Returns
+        -------
+        ht.tensor of the calculated means
+        """
+        mu_proc = __local_operation(torch.mean, x, out=None, **{'dim': dimen})
+        mu_tot_hold = tensor.zeros(target_dimens)
+        mu_tot = tensor.zeros(target_dimens)
+
+        beg = 0
+        if isinstance(dimen, (list, tuple, np.ndarray, tensor.tensor)):
+            # multiple dimensions
+            score = [True for itt in dimen if itt < x.split].count(True)
+        elif isinstance(dimen, (int, float)):
+            # one dimension
+            score = 1 if dimen < x.split else 0
+        for r in range(x.comm.Get_size()):
+            # need to add them together in order
+            if x.comm.rank == r:
+                end = beg + x.lshape[x.split]
+                if x.split - score == 0:
```

Comment:
what about arbitrary dimensions? you can compress this by explicitly calling __getitem__() with the proper slices filled with None:

`mu_tot_hold.__getitem__(x.split - score * [None] + [slice(beg, end)])`

(needs to be adjusted slightly maybe, from brain to terminal)

Reply:
this was removed as it is not needed in the new implementation (creation of a distributed heat tensor)